### PR TITLE
fix #489 when GlobalRollback but not receive report

### LINKED ISSUE #489: Data inconsistencies
- [x] I have searched the [issues](https://github.com/alibaba/fescar/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
I used 5 machines to test the data consistency of distributed transactions, and did some random network partition errors on 4 machines. After the test, I found that the data was inconsistent.

fescar version: 0.2.2 and using dubbo

|machine|process|
|-|-|
|172.28.128.2|nacos/nacos-server:latest, tester|
|172.28.128.3|samples-account,samples-storage,samples-order,samples-dubbo-business-call|
|172.28.128.4|samples-account,samples-storage,samples-order,samples-dubbo-business-call|
|172.28.128.5|samples-account,samples-storage,samples-order,samples-dubbo-business-call|
|172.28.128.6|samples-account,samples-storage,samples-order,samples-dubbo-business-call|

Random network partition in 172.28.128.3,172.28.128.4，172.28.128.5 and 172.28.128.6

The test program calls the data each time as below:
```java
Business business = new Business();
business.setAmount(new BigDecimal(1));
business.setCommodityCode(storage.getCommodityCode());
business.setUserId(account.getUserId());
business.setName(storage.getName());
business.setCount(1);
```

### Ⅱ. Describe what happened
Manually close the test program after running for a while，i  found that the data was inconsistent.

```sql

MySQL [db_gts_fescar]> select * from t_account;
+----+---------+---------+
| id | user_id | amount  |
+----+---------+---------+
|  1 | 1       | 3888.00 |
+----+---------+---------+
1 row in set (0.00 sec)

MySQL [db_gts_fescar]> select * from t_storage;
+----+----------------+--------+-------+
| id | commodity_code | name   | count |
+----+----------------+--------+-------+
|  1 | C201901140001  | 水杯   |  3890 |
+----+----------------+--------+-------+
1 row in set (0.00 sec)

MySQL [db_gts_fescar]> select * from t_order;
+-----+----------------------------------+---------+----------------+-------+--------+
| id  | order_no                         | user_id | commodity_code | count | amount |
+-----+----------------------------------+---------+----------------+-------+--------+
|  64 | 85d99e29d575441dbdc64c673199e0d8 | 1       | C201901140001  |     1 |   1.00 |
|  65 | 09ea4e3594784c2cb407e187144980e7 | 1       | C201901140001  |     1 |   1.00 |
|  66 | 21ce991a306843a69692d0266578873a | 1       | C201901140001  |     1 |   1.00 |
|  67 | 21947bd725ee4fed97306f2f3ecb764e | 1       | C201901140001  |     1 |   1.00 |
|  68 | 5e579dd278b14b21bebff549ddf0f875 | 1       | C201901140001  |     1 |   1.00 |
|  69 | 387adedfb5dd462d96d212ad7701d628 | 1       | C201901140001  |     1 |   1.00 |
|  70 | 6a0870bdddf04fcd8bc6eb1d72ab08ad | 1       | C201901140001  |     1 |   1.00 |
|  71 | 8cf7052d74704694b477c3112497e184 | 1       | C201901140001  |     1 |   1.00 |
|  72 | f19ad9293b3a4c62a86872af88a458f6 | 1       | C201901140001  |     1 |   1.00 |
|  73 | 6c766da501aa4b53b550001221d2e28d | 1       | C201901140001  |     1 |   1.00 |
|  74 | 0a3c44647d1b4e16b2d90e77f74af9e1 | 1       | C201901140001  |     1 |   1.00 |
|  75 | 2bda1a3872414525b4ef43a6c7f17ba9 | 1       | C201901140001  |     1 |   1.00 |
|  76 | 041f8c59cf7d4be598b7cdb268c92f84 | 1       | C201901140001  |     1 |   1.00 |
|  77 | 7c7130d3b59f48bbb08d20c60d8064f3 | 1       | C201901140001  |     1 |   1.00 |
|  78 | 34778c9547c646608d3a1c4adb4e3b46 | 1       | C201901140001  |     1 |   1.00 |
|  79 | 27ad7b36f0f0438e89bdde8df60792dc | 1       | C201901140001  |     1 |   1.00 |
|  80 | b4198ea32d594f05aac1c3c187a8420c | 1       | C201901140001  |     1 |   1.00 |
|  81 | 4743b66dbc1c4b10afd97b605a3485bd | 1       | C201901140001  |     1 |   1.00 |
|  82 | 7d9508a766f44000b99b416b20d6b493 | 1       | C201901140001  |     1 |   1.00 |
|  83 | 82759566708042aab49ad75a81c82345 | 1       | C201901140001  |     1 |   1.00 |
|  84 | 3900b0f634964c0c91defbf0dad98f0b | 1       | C201901140001  |     1 |   1.00 |
|  85 | be8f0850d4894992a4c0111a0d5d0d92 | 1       | C201901140001  |     1 |   1.00 |
|  86 | 60716b63658240728dff3723e18d017e | 1       | C201901140001  |     1 |   1.00 |
|  87 | 8a28e4b57baa43fb83e5d9a5ec9aab53 | 1       | C201901140001  |     1 |   1.00 |
|  88 | c5eed979802d402b9f33b6be18dd3b69 | 1       | C201901140001  |     1 |   1.00 |
|  89 | 6b783dc4532d4549a86f5c9d8c390ae3 | 1       | C201901140001  |     1 |   1.00 |
|  90 | da95f6bd49804fbda6debaae8af5170b | 1       | C201901140001  |     1 |   1.00 |
|  91 | 335ddb80554d45809c9b7c448159deda | 1       | C201901140001  |     1 |   1.00 |
|  92 | 0718215ad91e41a0844f2f935c8a9549 | 1       | C201901140001  |     1 |   1.00 |
|  93 | 465a352feee04d409873ac5f0fbd7295 | 1       | C201901140001  |     1 |   1.00 |
|  94 | 4352a7da2a5d4e57a702667701ad615b | 1       | C201901140001  |     1 |   1.00 |
|  95 | e2cbcf7b9c364dc98192d92edcd3e476 | 1       | C201901140001  |     1 |   1.00 |
|  96 | 8902c86a1ef74bf2bd85f4418afd4a24 | 1       | C201901140001  |     1 |   1.00 |
|  97 | f78ffaa087b1456bb5aeda480e69b5e8 | 1       | C201901140001  |     1 |   1.00 |
|  98 | bd3ae274b5c8489589188fa4f94eacfb | 1       | C201901140001  |     1 |   1.00 |
|  99 | 929c45af6a2f49eaaaacb29de9c3e7c9 | 1       | C201901140001  |     1 |   1.00 |
| 100 | 60547ca12f2d49ce9e8abc0b91d1ca64 | 1       | C201901140001  |     1 |   1.00 |
| 101 | 2bc98aef1bed48cab635b861a618a3d7 | 1       | C201901140001  |     1 |   1.00 |
| 102 | 8e7d738b30a84a5caeae2d893793a41f | 1       | C201901140001  |     1 |   1.00 |
| 103 | b1d06640c97a4ac69fac4f2603778dc5 | 1       | C201901140001  |     1 |   1.00 |
| 104 | 0b4aa247e9954b3781e9a669a30654ef | 1       | C201901140001  |     1 |   1.00 |
| 105 | 9499da00c07644b19fc4ef29de4a8946 | 1       | C201901140001  |     1 |   1.00 |
| 106 | 580d5ebd5544497cab7b4e0eca1df3cb | 1       | C201901140001  |     1 |   1.00 |
| 107 | e8ee548080c34353b3fd2f781a0e9a5b | 1       | C201901140001  |     1 |   1.00 |
| 108 | 7d8f1733b3bc46c5a12166b1b161e2be | 1       | C201901140001  |     1 |   1.00 |
| 109 | 35ad92bd9d9b4595878a715748df7258 | 1       | C201901140001  |     1 |   1.00 |
| 110 | addd7d6793034fc99955d4c2e052d2c0 | 1       | C201901140001  |     1 |   1.00 |
| 111 | 25f3ea73c30c4f40b8c14fb0c1833a48 | 1       | C201901140001  |     1 |   1.00 |
| 112 | 8395955f5c104774beefe1d765950e8c | 1       | C201901140001  |     1 |   1.00 |
| 113 | 3ec18fa53b154203a4f439bdff2d0a25 | 1       | C201901140001  |     1 |   1.00 |
| 114 | 21eb59052caf4e29895bbf2e1c135262 | 1       | C201901140001  |     1 |   1.00 |
| 115 | 428db32808354fe4b9b11fff6df739a2 | 1       | C201901140001  |     1 |   1.00 |
| 116 | e2eb7c182f804f7eba6e4049eaa8c16e | 1       | C201901140001  |     1 |   1.00 |
| 117 | 5e52dedfdf42410a9b5d1f23a3e8a67b | 1       | C201901140001  |     1 |   1.00 |
| 118 | c84f86b36a104f3dbbe02a74ce530c2c | 1       | C201901140001  |     1 |   1.00 |
| 119 | 62d51fbe99644dc6b032feb79c14ea2c | 1       | C201901140001  |     1 |   1.00 |
| 120 | 6591abf681134d12a86738499dbc20b2 | 1       | C201901140001  |     1 |   1.00 |
| 121 | 9b8ecd3d28954825b95a8069196dea01 | 1       | C201901140001  |     1 |   1.00 |
| 122 | 4219fc71bffd44f6a5dcd7d4c109ab8d | 1       | C201901140001  |     1 |   1.00 |
| 123 | caca1b0c4aa34c3b8dac28eaf51978a3 | 1       | C201901140001  |     1 |   1.00 |
| 124 | ede4f1e8f53d4c358674f26fd46335af | 1       | C201901140001  |     1 |   1.00 |
| 125 | f6f889bb63ef469ab31d72e3513f941b | 1       | C201901140001  |     1 |   1.00 |
| 126 | c7273e47e4604155a2cb449e1e1731f1 | 1       | C201901140001  |     1 |   1.00 |
| 127 | 54be17959a3849e6ad2c45ec41c0f844 | 1       | C201901140001  |     1 |   1.00 |
| 128 | 2518e5c1bbe54ff8ab3ad50870dfa134 | 1       | C201901140001  |     1 |   1.00 |
| 129 | 41492d3f239344ab916073aae85114ce | 1       | C201901140001  |     1 |   1.00 |
| 130 | ecbea0a2e201422b91346631d396299e | 1       | C201901140001  |     1 |   1.00 |
| 131 | f035ec57cf384f62b175e244e106630e | 1       | C201901140001  |     1 |   1.00 |
| 132 | 1d1f63833d1e40aebba1ed3284fe2246 | 1       | C201901140001  |     1 |   1.00 |
| 133 | e1dce7b8e552460ba336f5a2f6d6fa90 | 1       | C201901140001  |     1 |   1.00 |
| 134 | c6ef07493942419a88ee2e55eb3a46f9 | 1       | C201901140001  |     1 |   1.00 |
| 135 | 2c95f4828f2349acaae1a7453424121e | 1       | C201901140001  |     1 |   1.00 |
| 136 | c0d1c34ead4d4568bae4ee2b3e94a64a | 1       | C201901140001  |     1 |   1.00 |
| 137 | 3bfe8d128b6d431d9e2eae6f7205015a | 1       | C201901140001  |     1 |   1.00 |
| 138 | 6c30cb21189142cd811b7798dac638ff | 1       | C201901140001  |     1 |   1.00 |
| 139 | ce0afce1a78c45648c6801357c99da56 | 1       | C201901140001  |     1 |   1.00 |
| 140 | 629a57904d3441d8ad225821302c8ce6 | 1       | C201901140001  |     1 |   1.00 |
| 141 | fde3d1e512944f069eabaa47bd8b03fa | 1       | C201901140001  |     1 |   1.00 |
| 142 | cb561ef53cdd47e091ea21add1425738 | 1       | C201901140001  |     1 |   1.00 |
| 143 | 01e3ec29596b4d79aa9d734943a5098d | 1       | C201901140001  |     1 |   1.00 |
| 145 | e00270a0ed1a4052986a61311a8a8190 | 1       | C201901140001  |     1 |   1.00 |
| 146 | 2f28ac010cc2456685020c5d6a539894 | 1       | C201901140001  |     1 |   1.00 |
| 147 | 43b2158d90a64dc3aaf9c443c403f99c | 1       | C201901140001  |     1 |   1.00 |
| 148 | b580ae973b9a4de194fd310b0964f896 | 1       | C201901140001  |     1 |   1.00 |
| 149 | d3c9f4226cdf4a5c9e29997049614cf7 | 1       | C201901140001  |     1 |   1.00 |
| 150 | 392988002bed4d77a3f47cb5a788c727 | 1       | C201901140001  |     1 |   1.00 |
| 151 | 1f4f567d8a234f3882101b73f88b3b87 | 1       | C201901140001  |     1 |   1.00 |
| 152 | 87cbca20987b4e2bb4f5ff1a9cf7748d | 1       | C201901140001  |     1 |   1.00 |
| 153 | e2ed19253bd04070b8a20b7da2f906dd | 1       | C201901140001  |     1 |   1.00 |
| 154 | e11856f6c9e74380b59cdad4533b969a | 1       | C201901140001  |     1 |   1.00 |
| 155 | a1020c7618094db48192c7b7a043f4b0 | 1       | C201901140001  |     1 |   1.00 |
| 156 | 1d0350e8825f4a2ba173e77e30f1c002 | 1       | C201901140001  |     1 |   1.00 |
| 157 | 7ee815f928434f4ba748f2d0b30b657b | 1       | C201901140001  |     1 |   1.00 |
| 158 | d90a240a34b248208b7a2bab9f673c9f | 1       | C201901140001  |     1 |   1.00 |
| 159 | 425553ba24914055adbc24f14853d74a | 1       | C201901140001  |     1 |   1.00 |
| 160 | 783b292bd6264defb946e2d9a9755e0a | 1       | C201901140001  |     1 |   1.00 |
| 161 | af1bbe37ea144ac8a7ebd8031e34939b | 1       | C201901140001  |     1 |   1.00 |
| 162 | 66f3af9e31c64d3e80611ca7c50bf55b | 1       | C201901140001  |     1 |   1.00 |
| 163 | 9df3e3019db54fc0b4286c9945915754 | 1       | C201901140001  |     1 |   1.00 |
| 164 | 55cee59e927f4014bed0c352c4db81fc | 1       | C201901140001  |     1 |   1.00 |
| 165 | 145451ed1dcf447aa1c464ab7b04a5bd | 1       | C201901140001  |     1 |   1.00 |
| 167 | 1c1ad25486a641869f344172de1b66e9 | 1       | C201901140001  |     1 |   1.00 |
| 168 | 8b6d26f6e28d4efababd6bc5a821a81b | 1       | C201901140001  |     1 |   1.00 |
| 169 | 8f97199210984fc49ad6886194262d3d | 1       | C201901140001  |     1 |   1.00 |
| 170 | c1da36e73772422b8f6a0a3253bc2cab | 1       | C201901140001  |     1 |   1.00 |
| 171 | 3cf3490a100b4d218f6b3a34c6fd241f | 1       | C201901140001  |     1 |   1.00 |
| 172 | 7685a68ae6c74d988f501e748db3cdac | 1       | C201901140001  |     1 |   1.00 |
| 173 | 7a862a84441d4cc4bb9b3ebad257523a | 1       | C201901140001  |     1 |   1.00 |
| 174 | 129b34bfe6774f408dfb65fdca878738 | 1       | C201901140001  |     1 |   1.00 |
| 175 | b29b6cc9f699460b8bdfe765f98fd23a | 1       | C201901140001  |     1 |   1.00 |
| 176 | 37014df58e8d467ea8ce72e99a77b4ea | 1       | C201901140001  |     1 |   1.00 |
| 177 | adb94a345d014664a7ef1b69b110caba | 1       | C201901140001  |     1 |   1.00 |
+-----+----------------------------------+---------+----------------+-------+--------+
112 rows in set (0.00 sec)

MySQL [db_gts_fescar]> select * from undo_log;
Empty set (0.00 sec)
```
You can see that there are 112 successful orders, and the inventory and user balance data does not match, and there are no records in the undo_log table, all global transactions have been completed.

In the logs of the fescar-server, no errors were found, and the logs were successfully submitted.

### Ⅲ. Describe what you expected to happen
expect data:
1. t_account amount: 4000-112 = 3888
2. t_storage count: 4000 - 112 = 3888
3. t_order: 112 records
4. undo_log: no records

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- OS :
- Others:

### PR DESCRIPTION
**Pull request by zhangxu19830126**
### Ⅰ. Describe what this PR did
Fix #489 

### Ⅱ. Does this pull request fix one issue?

fixes #489 

### Ⅳ. Describe how to verify it
A, B two processes, perform the following steps:
1. A initiates a global transaction
2. B registration branch transaction succeeded
3. B service processing time is greater than the RPC timeout period.
4. A receives the RPC timeout and initiates a global transaction rollback.
5. After the global transaction is rolled back, the business of B is processed.

### CODE REVIEW

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java [hunk 1]
```diff
@@ -42,19 +41,40 @@
...
+        }
+
+        public int getValue() {
+            return value;
+        }
+    }
 
     private static final Logger LOGGER = LoggerFactory.getLogger(UndoLogManager.class);
 
     private static String UNDO_LOG_TABLE_NAME = "undo_log";
```
Comment (xingfudeshi) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: final?

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java [hunk 2]
```diff
@@ -42,19 +41,40 @@
...
 
     private static String INSERT_UNDO_LOG_SQL = "INSERT INTO " + UNDO_LOG_TABLE_NAME + "\n" +
```
Comment (xingfudeshi) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: final?

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java [hunk 3]
```diff
@@ -42,19 +41,40 @@
...
         "\t(branch_id, xid, rollback_info, log_status, log_created, log_modified)\n" +
-        "VALUES (?, ?, ?, 0, now(), now())";
+        "VALUES (?, ?, ?, ?, now(), now())";
     private static String DELETE_UNDO_LOG_SQL = "DELETE FROM " + UNDO_LOG_TABLE_NAME + "\n" +
```
Comment (xingfudeshi) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: final?

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java [hunk 4]
```diff
@@ -42,19 +41,40 @@
...
         "\tWHERE branch_id = ? AND xid = ?";
 
     private static String SELECT_UNDO_LOG_SQL = "SELECT * FROM " + UNDO_LOG_TABLE_NAME
```
Comment (xingfudeshi) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: final?

Comment (zhangxu19830126) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: yes, any problems?

Comment (xingfudeshi) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: > yes, any problems?

It is recommended that you use final keyword for a constant .

Comment (xingfudeshi) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: It is recommended that you use final keyword for a constant .

Comment (zhangxu19830126) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: This has nothing to do with this issue, and it is related to checkstyle. Does the checkstyle related problem need to be handled separately?

File: rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java [hunk 5]
```diff
@@ -125,70 +127,107 @@ public static void undo(DataSourceProxy dataSourceProxy, String xid, long branch
...
+                // causing undo_log not to be written to the database.
+                // For example, the business processing timeout, the global transaction is the initiator rolls back.
+                // To ensure data consistency, we can insert an undo_log with GlobalFinished state
+                // to prevent the local transaction of the first phase of other programs from being correctly submitted.
+                // See https://github.com/alibaba/fescar/issues/489
+
+                if (exists) {
+                    deleteUndoLog(xid, branchId, conn);
+                    conn.commit();
+                    LOGGER.info("xid {} branch {}, undo_log deleted",
```
Comment (sharajava) [inline rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: logging placeholder seems is missed for state.

Comment (zhangxu19830126) [reply rm-datasource/src/main/java/com/alibaba/fescar/rm/datasource/undo/UndoLogManager.java]: Yes，good job!

**Comment by skyesx** [comment]
does this solution handles the situation 
* TC invoke rollback 
*  and RM undo executed and log clean,
* and TC receive RM return timeout
* and TC call again

**Comment by slievrly** [comment]
LGTM

**Comment by slievrly** [comment]
> does this solution handles the situation
> 
> * TC invoke rollback
> * and RM undo executed and log clean,
> * and TC receive RM return timeout
> * and TC call again

1. TC invoke rollback but report timeout
2. when log clean ，just insert undolog to prevent local transactions from continuing
3. timeout->retry 
4. when insert fake undolog and execute real undo-> success

**Comment by skyesx** [comment]
> > does this solution handles the situation
> > 
> > * TC invoke rollback
> > * and RM undo executed and log clean,
> > * and TC receive RM return timeout
> > * and TC call again
> 
> 1. TC invoke rollback but report timeout
> 2. when log clean ，just insert undolog to prevent local transactions from continuing
> 3. timeout->retry
> 4. when insert fake undolog and execute real undo-> success

You are right

**Pull request merged**